### ISSUE TITLE
interactions between views + autograd.Function + AOTAutograd causes memory leak

### ISSUE THREAD

**Issue by bdhirsh** [pytorch/pytorch#94990]
I have a repro here (actively working on making it smaller). What's interesting is that it's non-trivial to make smaller - removing some extra ops, and e.g. changing the `autograd.function.apply()` to instead just call `torch.ops.test.foo_fwd()` directly all remove the leak.

```
import torch

def foo_fwd(v, o):
    out = torch.randn(512 * 256 * 128, device=v.device)
    return out.view(512, 256, 128)

def foo_bwd(grad_output, o):
    out = torch.zeros(64656, 128, device=grad_output.device)
    return out.squeeze(-1)

lib = torch.library.Library("test", "DEF")
lib.define("foo_fwd(Tensor v, Tensor[] o) -> Tensor")
lib.define("foo_bwd(Tensor grad_output, Tensor[] o) -> Tensor")

cuda_lib = torch.library.Library("test", "IMPL", "CUDA")
cuda_lib.impl("foo_fwd", foo_fwd)
cuda_lib.impl("foo_bwd", foo_bwd)

meta_lib = torch.library.Library("test", "IMPL", "Meta")
meta_lib.impl("foo_fwd", foo_fwd)
meta_lib.impl("foo_bwd", foo_bwd)

class Foo(torch.autograd.Function):
    @staticmethod
    def forward(ctx, v, o):
        return torch.ops.test.foo_fwd(v, o)

    @staticmethod
    def backward(ctx, grad_outputs):
        o = torch.ones(513, device='cuda', dtype=torch.int32)
        return torch.ops.test.foo_bwd(grad_outputs[0], [o])


def f(a, b, c, d, e, f):
    n: int = c.size(1)
    g = torch.mm(a, d)
    _, out, h = torch.split(g, [256, 256, 128], dim=1)
    i = h.unsqueeze(-1) * e + f
    j, _ = torch.unbind(i, dim=-1)
    k = Foo.apply(j, [b])
    torch._dynamo.graph_break()
    tmp = k + 1
    return out

f_compiled = torch.compile(f, backend="aot_eager")
inps = [
    torch.randn(64656, 256, dtype=torch.float32, device='cuda'),
    torch.arange(513, dtype=torch.int32, device='cuda'),
    torch.randn(512, 256, dtype=torch.float32, device='cuda'),
    torch.empty((256, 640), device='cuda', requires_grad=True),
    torch.empty(1, 128, 2, device='cuda', requires_grad=True),
    torch.zeros(1, 128, 2, device='cuda', requires_grad=True),
]

# This eventually OOMs
for i in range(300):
    print(torch.cuda.memory_allocated())
    #output = f(*inps)
    output = f_compiled(*inps)
    del output
    print(torch.cuda.memory_allocated())
#    print()
#print("@@@@@")
```

**Comment by bdhirsh** [comment]
Here's a slightly smaller repro (cut out the custom op and a few extra ops):
```
class Foo(torch.autograd.Function):
    @staticmethod
    def forward(ctx, v):
        return torch.randn(512, 256, 128, device=v.device)

    @staticmethod
    def backward(ctx, grad_outputs):
        return None

def f(a, d, e, f):
    g = a + d
    _, out, h = torch.split(g, [256, 256, 128], dim=1)
    i = h.unsqueeze(-1) * e + f
    j, _ = torch.unbind(i, dim=-1)
    k = Foo.apply(j)
    torch._dynamo.graph_break()
    tmp = k + 1
    return out

f_compiled = torch.compile(f, backend="aot_eager")
inps = [
    torch.randn(64656, 640, dtype=torch.float32, device='cuda'),
    torch.empty(64656, 640, device='cuda', requires_grad=True),
    torch.empty(1, 128, 2, device='cuda', requires_grad=True),
    torch.zeros(1, 128, 2, device='cuda', requires_grad=True),
]

# This eventually OOMs
for i in range(300):
    print(torch.cuda.memory_allocated())
    #output = f(*inps)
    output = f_compiled(*inps)
    del output
    print(torch.cuda.memory_allocated())
```

**Comment by ezyang** [comment]
I would not be surprised if this was leaking due to a reference cycle

**Comment by bdhirsh** [comment]
fwiw - I tried running the python gc manually at the end of the AOTAutograd epilogue, with no luck. But that theory still seems valid.

**Comment by bdhirsh** [comment]
update: custom `autograd.Function` not actually necessary (only the graph break + a bunch of interesting views is needed to repro):
```
import torch

def f(a, d, e, f):
    g = a + d
    _, out, h = torch.split(g, [256, 256, 128], dim=1)
    i = h.unsqueeze(-1) * e + f
    j, _ = torch.unbind(i, dim=-1)
    torch._dynamo.graph_break()
    k = j + 1
    return out

f_compiled = torch.compile(f, backend="aot_eager")
inps = [
    torch.randn(64656, 640, dtype=torch.float32, device='cuda'),
    torch.empty(64656, 640, device='cuda', requires_grad=True),
    torch.empty(1, 128, 2, device='cuda', requires_grad=True),
    torch.zeros(1, 128, 2, device='cuda', requires_grad=True),
]

# This eventually OOMs
for i in range(300):
    print(torch.cuda.memory_allocated())
    #output = f(*inps)
    output = f_compiled(*inps)
    del output
    print(torch.cuda.memory_allocated())
```

**Comment by bdhirsh** [comment]
Smaller repro that cuts out dynamo completely (this one only uses `aot_function`):
```
import torch
from functorch.compile import aot_function, nop

inps = [
    torch.ones(64656, 640, device='cuda', requires_grad=True),
    torch.ones(1, 128, 2, device='cuda', requires_grad=True) * 5,
]

def graph(a, b):
    add = a + a
    split = torch.functional.split(add, [256, 256, 128], dim=1)
    getitem_2 = split[2]
    unsqueeze = getitem_2.unsqueeze(-1)
    mul = unsqueeze * b
    return (getitem_2, mul)

f_compiled = aot_function(
    graph,
    fw_compiler=nop,
    bw_compiler=nop,
)

# This eventually OOMs
for i in range(300):
    output = f_compiled(*inps)
    del output
    print(torch.cuda.memory_allocated())
```

**Comment by ezyang** [comment]
so, fwiw, if you're accidentally doing a `copy_` in the functionalized graph when you didn't need to, this could easily result in a history leak (because you are continuing to accumulate history on inps[1], which is not a leaf and thus can be inplace modified.

**Comment by bdhirsh** [comment]
Good point - I don't think I am, but I can double check.

**Comment by ezyang** [comment]
grad_fn stays MulBackward so it's probably not this

**Comment by bdhirsh** [comment]
I removed `AOTAutograd` from the repro: now it's only using `autograd.Function`:

```
import torch

inps = [
    torch.ones(64656, 640, device='cuda', requires_grad=True),
    torch.ones(1, 128, 2, device='cuda', requires_grad=True) * 5,
]

def f(a, b):
    add = torch.ops.aten.add.Tensor(a, a)
    split_with_sizes = torch.ops.aten.split_with_sizes.default(add, [256, 256, 128], 1)
    getitem_2 = split_with_sizes[2]
    unsqueeze = torch.ops.aten.unsqueeze.default(getitem_2, -1)
    mul = torch.ops.aten.mul.Tensor(unsqueeze, b)
    return [getitem_2, mul, add, b, unsqueeze]

class TestFn(torch.autograd.Function):
    @staticmethod
    def forward(ctx, *inps):
        num_forward_returns = 3
        getitem_2, mul, add, b, unsqueeze = f(*inps)
        ctx.save_for_backward(unsqueeze)
        return getitem_2, mul, add

    @staticmethod
    def backward(ctx, *flat_args):
        return None

# This eventually OOMs
for i in range(300):
    output = TestFn.apply(*inps)
    del output
    print(torch.cuda.memory_allocated())
```

Interesting thing to note: the leak goes away when I don't save `unsqueeze` for backward.

**Comment by bdhirsh** [comment]
new smallest repro:

```
import torch

inps = [
    torch.ones(64656, 640, device='cuda', requires_grad=True),
]

def f(a):
    b = a + 1
    c = b[0]
    d = c.unsqueeze(-1)
    return [b, c, d]

class TestFn(torch.autograd.Function):
    @staticmethod
    def forward(ctx, a):
        b, c, d = f(a)
        ctx.save_for_backward(d)
        return b, c

    @staticmethod
    def backward(ctx, *flat_args):
        return None

# This eventually OOMs
for i in range(300):
    output = TestFn.apply(*inps)
    del output
    print(torch.cuda.memory_allocated())
```

**Comment by bdhirsh** [comment]
A few updates. First, a minimal repro:
```
import torch
import objgraph

a = torch.ones(64656, 640, device='cuda', requires_grad=True)

class TestFn(torch.autograd.Function):
    @staticmethod
    def forward(ctx, a):
        b = a + 1
        c = b.view(-1)
        ctx.save_for_backward(c)
        return b

    @staticmethod
    def backward(ctx, *flat_args):
        return None

print(len(objgraph.by_type('torch.Tensor')))
TestFn.apply(a)
print(len(objgraph.by_type('torch.Tensor')))
TestFn.apply(a)
print(len(objgraph.by_type('torch.Tensor')))
```

The above prints "1, 3, 5": after each call to the `autograd.Function, we're leaking two extra tensors.

I talked to Alban about the memory leak. Summarizing my discussion with him:

**(1) root case (ref cycle)**

The root cause is a reference cycle between C++ and python state:
<img width="1059" alt="image" src="https://user-images.githubusercontent.com/16311747/219779218-f829b65c-9af9-4dbc-ac5c-1d49bd25adc5.png">



**(2) Possible fixes**

We need to break the ref cycle in some way. Some options:

(a) A "good" fix would be to make `tp_traverse()` smarter (in `python_variable.cpp`), so that the python gc can automatically detect this ref cycle. There's already precedent for this: our custom `THPVariable_subclass_traverse()` already knows to potentially traverse the `AutogradMeta` of a tensor ([code](https://github.com/pytorch/pytorch/blob/e0ede1cc309a2cc11f8f84b1823e605f9fd15c4b/torch/csrc/autograd/python_variable.cpp#L2007)) - it just doesn't give the same treatment to `ViewAutogradMeta`.

(b) a lower-blast-radius change would be to try to directly break the ref cycle. One way to do this would be to break the cycle between (3) and (4) in the above diagram, by saving `c.detach()` for backward instead of `c`. Autograd does a similar trick today for when a saved-for-backward variable is also one of the outputs to the autograd.Function.

**Why did this leak start showing up after https://github.com/pytorch/pytorch/pull/92076**

That PR basically causes `AOTAutograd` to check for "outputs that alias intermediate bases" in each graph, and directly return those intermediate bases as outputs in the `autograd.Function`. If we happened to be saving the alias for backward, then the act of returning the intermediate base induces the leak.

**Next steps?**

@albanD can correct me if anything I put sounds wrong. One question here is on how critical it is to fix this bug for the release. It sounds to me like (a) is a nice longer term solution, while something similar to (b) might be lower blast radius and easier to verify.

I would say that we want this fix in for the release, if it causes an OOM on 1 model out of the 100+ on torchbench. But I'm not strongly tied to that opinion / curious what others think.

**Comment by ezyang** [comment]
Is there a known problem with (b)?

**Comment by bdhirsh** [comment]
@albanD also pointed out a quick 1-liner that should fix the problem for AOTAutograd (but doesn't fix the underlying autograd leak:

```
ctx.save_for_backward(map(lambda x: x.detach(), to_save))
```

Basically, force-detach any views that we save for backward. This will be completely wrong for double backward, but is fine as long as we're not doing double backward (which AOTAUtograd explicitly errors on today anyway)

**Comment by ezyang** [comment]
None of these seem relevant to AOTAutograd. Ship it!

**Comment by bdhirsh** [comment]
Are you ok with the 1-liner for the release? I just confirmed that it fixes the OOM on the internal model. The only real downside that I see is we need to construct num_saved_variables calls to `.detach()` in the hotpath, but I think we do more egregious things in the AOTAutograd hotpath today.

I think (b) is still pretty non-trivial to implement (I don't know exactly how to do it offhand, but it seems possible)

**Comment by bdhirsh** [comment]
It's still possible to encounter this ref cycle when constructing an `autograd.Function`, but this has been worked around in AOTAutograd, so it isn't really a problem for `torch.compile`. Removing the hi-pri label. (there was a draft PR to try to fix the ref cycle that ran into some issues here: https://github.com/pytorch/pytorch/pull/95904)
